core/trino-main/src/main/java/io/trino/operator/InterpretedHashGenerator.java

```diff
@@ -95,13 +135,74 @@ private long nullSafeHash(int operatorIndex, Block block, int position)
+                    }
+                }
+                else {
+                    ValueBlock valueBlock = dictionaryBlock.getUnderlyingValueBlock();
+                    for (int i = 0; i < length; i++) {
+                        hashes[i] = nullSafeHash(operatorIndex, valueBlock, dictionaryBlock.getUnderlyingValuePosition(i + positionOffset));
+                    }
+                }
+            }
+            case ValueBlock valueBlock -> nullSafeHash(operatorIndex, valueBlock, positionOffset, length, hashes);
+        }
+    }
+
+    private void hashBlockWithCombine(Block rawBlock, int positionOffset, int length, int operatorIndex, long[] hashes)
+    {
+        switch (rawBlock) {
+            case RunLengthEncodedBlock rleBlock -> {
+                long hash = nullSafeHash(operatorIndex, rleBlock.getUnderlyingValueBlock(), 0);
+                CombineHashFunction.combineAllHashesWithConstant(hashes, 0, length, hash);
+            }
+            case DictionaryBlock dictionaryBlock -> {
+                if (isDictionaryProcessingFaster(dictionaryBlock, length)) {
+                    ValueBlock dictionary = dictionaryBlock.getDictionary();
+                    long[] dictionaryHashes = new long[dictionary.getPositionCount()];
+                    nullSafeHash(operatorIndex, dictionary, 0, dictionary.getPositionCount(), dictionaryHashes);
+                    for (int i = 0; i < length; i++) {
+                        long hash = dictionaryHashes[dictionaryBlock.getId(i + positionOffset)];
+                        hashes[i] = CombineHashFunction.getHash(hashes[i], hash);
+                    }
+                }
+                else {
+                    ValueBlock valueBlock = dictionaryBlock.getUnderlyingValueBlock();
+                    for (int i = 0; i < length; i++) {
+                        long hash = nullSafeHash(operatorIndex, valueBlock, dictionaryBlock.getUnderlyingValuePosition(i + positionOffset));
+                        hashes[i] = CombineHashFunction.getHash(hashes[i], hash);
+                    }
+                }
+            }
+            case ValueBlock valueBlock -> {
+                long[] valueHashes = new long[length];
+                nullSafeHash(operatorIndex, valueBlock, positionOffset, length, valueHashes);
+                for (int i = 0; i < length; i++) {
+                    hashes[i] = CombineHashFunction.getHash(hashes[i], valueHashes[i]);
+                }
+            }
+        }
+    }
+
+    private static boolean isDictionaryProcessingFaster(DictionaryBlock dictionaryBlock, int length)
+    {
+        // if the dictionary block length is greater than the number of elements in the dictionary,
+        // it will be faster to compute hash for the dictionary values only once and re-use it
+        // instead of recalculating it.
+        return length > dictionaryBlock.getDictionary().getPositionCount();
```

Comment:
I'm betting the actual threshold here is more like `length >= (dictionaryBlock.getDictionary().getPositionCount() * 2)` since the overheads associated with this strategy are non-trivial.

Reply:
I've updated this to `length > dictionaryBlock.getDictionary().getPositionCount() * 1.2`